drivers/sensor/ina700/ina700.c

```diff
@@ -0,0 +1,333 @@
+		temp = FIELD_GET(IAN700_DIE_TEMPERATURE, data->temperature);
+		temp *= INA700_TEMPERATURE_LSB; /* m°C */
+		val->val1 = temp / 1000L;
+		val->val2 = temp % 1000L;
+		break;
+
+	default:
+		rc = -ENOTSUP;
+		break;
+	}
+
+	ina700_unlock_data(dev);
+
+	return 0;
+}
+
+#ifdef CONFIG_INA700_TRIGGER
+
+static void ina700_lock_data(const struct device *dev)
+{
+	struct ina700_data *data = dev->data;
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+}
+
+static void ina700_unlock_data(const struct device *dev)
+{
+	struct ina700_data *data = dev->data;
+
+	k_mutex_unlock(&data->mutex);
+}
+
+static void ina700_gpio_callback(const struct device *port, struct gpio_callback *cb, uint32_t pins)
+{
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+
+	struct ina700_data *data = CONTAINER_OF(cb, struct ina700_data, gpio_cb);
+
+	k_work_submit(&data->work);
+}
+
+static void ina700_work_cb(struct k_work *work)
+{
+	uint16_t status = 0;
+	struct ina700_data *data = CONTAINER_OF(work, struct ina700_data, work);
+
+	ina700_lock_data(data->this);
+
+	if (reg_read_16(data->this, INA700_REG_ALERT_DIAGNOSTICS, &status) == 0) {
+		if (status & INA700_ALERT_DIAG_CONVERSION_READY) {
+			if (ina700_sample_fetch(data->this, SENSOR_CHAN_ALL) == 0) {
+				if (data->handler_alert) {
+					data->handler_alert(data->this, data->trigger_alert);
+				}
+			}
+		}
+	}
+
+	ina700_unlock_data(data->this);
+}
+
+int ina700_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+		       sensor_trigger_handler_t handler)
+{
+	struct ina700_data *data = dev->data;
+
+	data->handler_alert = handler;
+	data->trigger_alert = trig;
+
+	return 0;
+}
+
+int ina700_trigger_init(const struct device *dev)
+{
+	int rc = 0;
+	struct ina700_data *data = dev->data;
+	const struct ina700_config *cfg = dev->config;
+
+	k_mutex_init(&data->mutex);
+	data->this = dev; /* for the trigger callback */
+	data->work.handler = ina700_work_cb;
+
+	if (!gpio_is_ready_dt(&cfg->alert_gpio)) {
+		LOG_ERR("Alert GPIO device not ready");
+		return -ENODEV;
+	}
+
+	rc = gpio_pin_configure_dt(&cfg->alert_gpio, GPIO_INPUT);
+
+	if (rc < 0) {
+		LOG_ERR("Failed to configure alert GPIO pin");
+		return rc;
+	}
+
+	gpio_init_callback(&data->gpio_cb, ina700_gpio_callback, BIT(cfg->alert_gpio.pin));
+	rc = gpio_add_callback(cfg->alert_gpio.port, &data->gpio_cb);
+
+	if (rc < 0) {
+		LOG_ERR("Failed to add alert GPIO callback");
+		return rc;
+	}
+
+	rc = gpio_pin_interrupt_configure_dt(&cfg->alert_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+
+	if (rc < 0) {
+		LOG_ERR("Failed to configure alert GPIO interrupt");
+		return rc;
+	}
+
+	/* only trigger on conversion ready, after averaging is done */
+	uint16_t alert_config =
+		INA700_ALERT_DIAG_CONVERSION_ALERT_ENABLE | INA700_ALERT_DIAG_SLOW_ALERT_MODE;
+
+	return reg_write_16(dev, INA700_REG_ALERT_DIAGNOSTICS, alert_config);
+}
+
+#else
+
+static void ina700_lock_data(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+}
+
+static void ina700_unlock_data(const struct device *dev)
```

Comment:
I think you can drop these wrappers and use mutex_lock/mutex_unlock directly within functions :)

Reply:
Then we would need to wrap each call to mutex_lock/unlock in an ifdef block - the mutex member is only defined if the trigger is enabled through kconfig.

Reply:
Yeah but why not use mutex each time you access the data, probably its not done in all the drivers, but i think it would be completely alright. 

Or atleast have just one function definition and you can do ifdefery within that function :)